### PR TITLE
Pidgin fixes and additional protocols

### PR DESCRIPTION
**Pull request by grimmy**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

Hi, I am the lead developer of [pidgin instant messenger](https://pidgin.im).  A user asked about how to install [purple-facebook](https://github.com/dequis/purple-facebook) with pidgin installed via homebrew, and rather than walk them through all of the arguments to configure, I decided it would be easier if I just packaged it.

No tests were run, because the existing formula does not have any and there isn't really a good way to test a GUI application.  That said, we could maybe add a version check?

However, when I went to do that, I discovered that pidgin was looking for plugins in a place that other formulas should not be installing files.  That is in `/usr/local/Cellar/pidgin/{VERSION}/lib/purple-2`.  So to address that, I updated the pidgin formula to 2.13.0_4 and used `inreplace` to change the directories that pidgin is looking for plugins.

By default pidgin just uses its `${PREFIX}`, but the due to the way that the homebrew tree is structured, it causes the above mentioned issue.  By instead replacing that code with the path `/usr/local/lib/purple-2` other formulas can just be installed and linked as normal and everything just works.

I've also included formulas for a lot of the new/modern instant messaging networks, however there's quite a bit of `inreplace`'s as their build systems are geared towards making things easier for users at the expense of packagers.  Also, absolutely none of these plugins have released tarballs, so they are all head only formulas.  I've left them requiring `--HEAD` to install, not sure if that was the right way to go or not.

### CODE REVIEW

File: Formula/purple-discord.rb
Comment (MikeMcQuaid) [inline Formula/purple-discord.rb:4]: I'm afraid we won't accept head-only formulae in Homebrew/core.

File: Formula/pidgin.rb
Comment (MikeMcQuaid) [inline Formula/pidgin.rb:57]: Could these be `configure` arguments in future?

Comment (grimmy) [reply Formula/pidgin.rb:57]: The problem is the symlinking in the homebrew tree.  The configure argument is `--prefix` and pidgin is looking there by default.  It's only the linking phase that homebrew does that causes this not to work and as such doesn't seem like it should be an upstream fix to me.

For example, `--prefix` is `/usr/local/Cellar/pidgin/2.13.0/` so pidgin/libpurple look in `/usr/local/Cellar/pidgin/2.13.0/lib/purple-2` which works for everything in tree and why I suspect that the otr plugin as added into the main formula as well.

However, formula's shouldn't be installing files to another formula's cellar path, so we have to convince libpurple to ignore it's prefix and instead look in another place (`/usr/local/lib/purple-2`) where other formulas can install files for their accompanying plugins.

File: Formula/purple-discord.rb
Comment (grimmy) [reply Formula/purple-discord.rb:4]: So the only way is with a separate tap?  Surely there's other projects/plugins out there that aren't cutting release tarballs?

Comment (MikeMcQuaid) [reply Formula/purple-discord.rb:4]: > Surely there's other projects/plugins out there that aren't cutting release tarballs?

Yup but we don't accept those projects' inclusion into Homebrew/core. If they are HEAD there's no difference to being in a tap; you aren't getting a binary package anyway (and it means we don't get issues if upstream breaks compilation).

File: Formula/pidgin.rb
Comment (MikeMcQuaid) [reply Formula/pidgin.rb:57]: > as such doesn't seem like it should be an upstream fix to me

I don't have to be but given Homebrew is a pretty widely used downstream I don't trust us (or one of our contributors) to not mess this up in future. The benefit of `configure` arguments is that you, upstream, can decide on how they work rather than us randomly patching files (where we'll be guaranteed to miss some on a version update).

To be clear: this is not blocking but just a "nice to have" I mention as you're upstream.

**Comment by MikeMcQuaid** [comment]
> No tests were run, because the existing formula does not have any and there isn't really a good way to test a GUI application. That said, we could maybe add a version check?

Yeh, that'd be preferable to nothing 👍

File: Formula/purple-discord.rb
Comment (MikeMcQuaid) [reply Formula/purple-discord.rb:4]: An option could be pinning to a specific `revision` and making them all `resource`s inside the main `pidgin` formula (so it comes with these by default).

File: Formula/pidgin.rb
Comment (grimmy) [reply Formula/pidgin.rb:57]: It won't happen upstream.  Pidgin 2 has one more minor release in it before it goes into feature freeze forever.  We have minimal resources and need to point that at our next major revision that's been slowly being worked on for a decade and thus pidgin 2 will be entering maintenance mode at that point.

File: Formula/purple-discord.rb
Comment (grimmy) [reply Formula/purple-discord.rb:4]: So to be clear.  If these packages had tags that we could pull tarballs for, would that be enough for inclusion?

Comment (grimmy) [reply Formula/purple-discord.rb:4]: Uh.. that's not how we do things.  Also quite of few of these are the result of reverse engineering network protocols and as such, can change at any time, and need to be out of tree so that they can be quickly fixed and not force an update to the core application that might not be ready for release.

In other words, modular components are better in this case.

**Comment by kgraefe** [comment]
I wrote [a plugin](https://github.com/kgraefe/purple-add-prefix) specifically for that use-case [and tried to get it installed in Homebrew by default](https://github.com/Homebrew/homebrew-core/pull/42864). But it got rejected back then. :-(

I am maintaining my [own tap for Pidgin plugins](https://github.com/kgraefe/homebrew-pidgin-plugins) but unfortunately installation is a bit cumbersome now.

**Comment by MikeMcQuaid** [comment]
> I wrote [a plugin](https://github.com/kgraefe/purple-add-prefix) specifically for that use-case [and tried to get it installed in Homebrew by default](https://github.com/Homebrew/homebrew-core/pull/42864). But it got rejected back then. :-(

It got rejected for essentially the same reasons I had above: this code should live in the upstream to avoid us breaking it in future.



> I am maintaining my [own tap for Pidgin plugins](https://github.com/kgraefe/homebrew-pidgin-plugins) but unfortunately installation is a bit cumbersome now.

This may be a good fit for adding to our documentation and seeing if the other formulae can be included.

**Comment by grimmy** [comment]
> > I wrote [a plugin](https://github.com/kgraefe/purple-add-prefix) specifically for that use-case [and tried to get it installed in Homebrew by default](https://github.com/Homebrew/homebrew-core/pull/42864). But it got rejected back then. :-(
> 
> It got rejected for essentially the same reasons I had above: this code should live in the upstream to avoid us breaking it in future.

This doesn't belong in upstream.  Creating a LFS like file system tree outside of an applications prefix is outside of the scope of the upstream application.

> 
> > I am maintaining my [own tap for Pidgin plugins](https://github.com/kgraefe/homebrew-pidgin-plugins) but unfortunately installation is a bit cumbersome now.
> 
> This may be a good fit for adding to our documentation and seeing if the other formulae can be included.

If it comes down to it, I'll just create our own tap including everything because it sure seems like we're not going to be able to come to a consensus here.

File: Formula/purple-discord.rb
Comment (MikeMcQuaid) [reply Formula/purple-discord.rb:4]: Fine with me. My suggestion on "batteries included" is based on how we (Homebrew) do things and what will result in a better experience for our users. I am happy to compromise if these formulae can have stable release tarballs made.

File: Formula/pidgin.rb
Comment (MikeMcQuaid) [reply Formula/pidgin.rb:57]: In that case: could this be something that could be included in Pidgin 3? I am suggesting `configure` flags to set `LIBDIR` and `DATADIR` rather than indefinitely patching.

**Comment by MikeMcQuaid** [comment]
> This doesn't belong in upstream. Creating a LFS like file system tree outside of an applications prefix is outside of the scope of the upstream application.

You are entitled to your opinion of what belongs in upstream and what does. Given we're a macOS package manager used by millions of people: I'm entitled to my opinion of what should and shouldn't be in the formula. 

Having a couple of `inreplace` lines will break with future releases and will be incomplete with future changes. I would like to find some solution to this problem.

> If it comes down to it, I'll just create our own tap including everything because it sure seems like we're not going to be able to come to a consensus here.

I am happy to compromise and reach consensus. I've suggested several areas of compromise above. What things are you willing to compromise on to come to a consensus?

**Comment by grimmy** [comment]
Okay, lets really get all the cards on the table before I spend any more time on this.

# Pidgin

For Pidgin, it is doing 100% everything that a GNU Autotools based package should be doing and has been fine in nearly every distribution that I'm aware of for over 20 years.  Of course, there is always an exception to every rule and in this case it's [GNU/Guix](https://en.wikipedia.org/wiki/GNU_Guix).   They patch Pidgin as part of their packaging to deal with the issues brought on by their packaging solution. See the patch [here](https://git.savannah.gnu.org/cgit/guix.git/tree/gnu/packages/patches/pidgin-add-search-path.patch).

So what are our other options:

We can't pass `--libdir` to Pidgin's `configure` because that would break the encapsulation of the prefix on Homebrew.  For example, everything for Pidgin would be installed into `/usr/local/Cellar/pidgin-2.13.0/` except for `/lib` which would end up in `/usr/local/lib`.  So that's out as it breaks the very model of Homebrew.

We _could_ do what you're suggesting and special case Homebrew, but now we (Pidgin upstream) are carrying around a patch that's specific to one package manager and in reality will never be tested after initial adoption.  This moves where the breakage will happen, not that fact that it will break. So when the build is broken over time, whoever is updating the Homebrew formula now has to get Pidgin upstream involved and due to the current policy, will require a new Pidgin release.  This is doable, but this then also means that every other package manager will be updating for a new release as well.  So this policy pushes all of the work off to literally everyone else except for Homebrew which is where the breakage is, due to the Homebrew special case.

I am concerned that this is a code smell.  We (Pidgin upstream) should not have to care about how a package manager installed our software, If someone wants to put everything under `/opt/pidgin` via `--prefix` that's fine.  Or if they want to go crazy and do something like `--prefix=/pidgin --datadir=/pidgin-data --libdir=/usr/local/lib/pidgin`, that's fine as well it's all supported by GNU Autotools.  What we refuse to do is support is a leaky abstraction from one of the many package managers that handle our software.

Finally we could do as I've already done in this PR and patch the source code to deal with the Homebrew specific situation caused by `brew link`.  The worry that "Upstream could break Homebrew" is valid, but we won't.  We're doing what every single GNU Autotools package has done for a very long time.  And even if we did somehow cause a breakage, it can all be handled by the person working on the formula update and they don't need us to release an update to fix it.  Which then means that every other package manager isn't getting back to back versions either.  On top of that, if something were to change in Homebrew, it can also be fixed without our intervention immediately by the person updating the formula.

That said, wouldn't those breakages be handled by the CI preformed here?  I haven't done a lot of Homebrew packaging, but I'd suspect if a patch fails, then the build fails, and then that build failure would cause the CI to fail, which would then cause the PR to be blocked, and thus never break users, right?

# Plugins

Do we agree that, for the plugins to be included in the homebrew-core tap, that they must specific a URL that can be a github repository with a version attribute of a specific revision or tag?  An example of this is from @kgraefe's tap and can be seen below:

```ruby
class PurpleIcyque < Formula
  ...
  url "https://github.com/EionRobb/icyque.git",
    :revision => "ebc522743e98a6b9e1fa0703703fa5298862b60a"
  ...
end
```

# Closing

I hope this discussion has swayed your opinion as our end goal here is to make software available to users.

**Comment by MikeMcQuaid** [comment]
> We can't pass `--libdir` to Pidgin's `configure` because that would break the encapsulation of the prefix on Homebrew. For example, everything for Pidgin would be installed into `/usr/local/Cellar/pidgin-2.13.0/` except for `/lib` which would end up in `/usr/local/lib`.

Makes sense. I was thinking about a `--plugindir` or similar?

> That said, wouldn't those breakages be handled by the CI preformed here? I haven't done a lot of Homebrew packaging, but I'd suspect if a patch fails, then the build fails, and then that build failure would cause the CI to fail, which would then cause the PR to be blocked, and thus never break users, right?

Correct.

> The worry that "Upstream could break Homebrew" is valid, but we won't.

My concern is more: that additional places `LIBDIR` or `DATADIR` need to be replaced are added but we don't `inreplace` them so `pidgin` on Homebrew is broken in unexpected ways.

If you are not concerned about this, though, I'm prepared to let it slide given we have a comment linking to the discussion here.

> Do we agree that, for the plugins to be included in the homebrew-core tap, that they must specific a URL that can be a github repository with a version attribute of a specific revision or tag? 

It needs to be a tag _and_ revision and not just revision. This is a long-standing, documented policy: https://docs.brew.sh/Acceptable-Formulae#stable-versions


---

In summary: I do not consider this PR blocked on the `inreplace` changes. I do consider it blocked on adding HEAD-only formulae. If you can extract the plugin formulae into another PR this can be merged with a single, comment change (that I'll suggest now).

File: Formula/pidgin.rb
Comment (MikeMcQuaid) [inline Formula/pidgin.rb:56]:
```suggestion
    # patch pidgin to read plugins and allow them to live in separate formulae which can 
    # all install their symlinks into these directories. See: 
    #   https://github.com/Homebrew/homebrew-core/pull/53557

```

**Comment by grimmy** [comment]
> > We can't pass `--libdir` to Pidgin's `configure` because that would break the encapsulation of the prefix on Homebrew. For example, everything for Pidgin would be installed into `/usr/local/Cellar/pidgin-2.13.0/` except for `/lib` which would end up in `/usr/local/lib`.
> 
> Makes sense. I was thinking about a `--plugindir` or similar?

Our plugin directory has always been `${libdir}/purple-2` and needs to match between the application and the build system for obvious reasons.
 
> > The worry that "Upstream could break Homebrew" is valid, but we won't.
> 
> My concern is more: that additional places `LIBDIR` or `DATADIR` need to be replaced are added but we don't `inreplace` them so `pidgin` on Homebrew is broken in unexpected ways.
> 
> If you are not concerned about this, though, I'm prepared to let it slide given we have a comment linking to the discussion here.

Pidgin 2 will be hitting hard maintenance mode soon and will not have any additional minor releases after that point.  So yes there is no chance this will change.

> > Do we agree that, for the plugins to be included in the homebrew-core tap, that they must specific a URL that can be a github repository with a version attribute of a specific revision or tag?
> 
> It needs to be a tag _and_ revision and not just revision. This is a long-standing, documented policy: https://docs.brew.sh/Acceptable-Formulae#stable-versions

Gotcha, somehow I completely missed that part in the docs.

> In summary: I do not consider this PR blocked on the `inreplace` changes. I do consider it blocked on adding HEAD-only formulae. If you can extract the plugin formulae into another PR this can be merged with a single, comment change (that I'll suggest now).

Sounds good.  I've been working with the plugin maintainers to get their stuff tagged up, but haven't gotten it to happen yet.

I'm really not all that proficient with github... What am I doing with the suggestion here?  Just hitting commit and marking as resolved?

Also, I am far from a git history rewriting expert so it'll probably take me a bit to cherry pick out these two commits. But I think I can handle it.

**Comment by MikeMcQuaid** [comment]
> I'm really not all that proficient with github... What am I doing with the suggestion here? Just hitting commit and marking as resolved?

Yep.



> Also, I am far from a git history rewriting expert so it'll probably take me a bit to cherry pick out these two commits. But I think I can handle it.

We can handle it for you if that'd help.

**Comment by grimmy** [comment]
> > I'm really not all that proficient with github... What am I doing with the suggestion here? Just hitting commit and marking as resolved?
> 
> Yep.

Cool

> > Also, I am far from a git history rewriting expert so it'll probably take me a bit to cherry pick out these two commits. But I think I can handle it.
> 
> We can handle it for you if that'd help.

Just finished think I got it.

**Comment by grimmy** [comment]
Just wondering if there's anything left for me to do that is blocking this?

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
Thanks so much for your contribution (and patience)! Without people like you submitting PRs we couldn't run this project. You rock, @grimmy!

**Comment by jonchang** [comment]
Cancelled the publish run since this needs to be squashed and rebased to use `brew pr-publish`. Doing the rebase locally.

**Comment by MikeMcQuaid** [comment]
Thanks @jonchang, missed the second commit 🤦

**Pull request closed without merging**
